use-cases/eurac/hpo.py

```diff
@@ -0,0 +1,231 @@
+        run_config = train.RunConfig(
+            name="Eurac-Ray-Experiment",
+            stop={"training_iteration": args.max_iterations}
+        )
+
+        # Set resource allocation for each trial (number of GPUs and/or number of CPUs)
+        resources_per_trial = {"gpu": args.ngpus}
+
+        # Set up Ray Tune Tuner
+        tuner = tune.Tuner(
+            tune.with_resources(run_trial, resources=resources_per_trial),
+            tune_config=tune_config,
+            run_config=run_config,
+            param_space=search_space  # Search space defined above
+        )
+
+        # Run the hyperparameter optimization and get results
+        result_grid = tuner.fit()
+
+    else:
+        # Load results from an earlier Ray Tune run
+        print(f"Loading results from {args.experiment_path}...")
+
+        # Restore tuner from saved results
+        restored_tuner = tune.Tuner.restore(
+            args.experiment_path,
+            trainable=run_trial
+        )
+        result_grid = restored_tuner.get_results()
+
+    # Display experiment statistics
+    print(f"Number of errored trials: {result_grid.num_errors}")
+    print(f"Number of terminated trials: {result_grid.num_terminated}")
+    print(f"Ray Tune experiment path: {result_grid.experiment_path}")
+
+    # Get the best result based on the last 10 iterations' average
+    best_result = result_grid.get_best_result(
+        scope="last-10-avg",
+        metric=args.metric,
+        mode="min"
+    )
+    print(f"Best result: {best_result}")
+
+    # Print a dataframe with all trial results
+    result_df = result_grid.get_dataframe()
+    print(f"All results dataframe: {result_df}")
+    print(f"All result columns: {result_df.columns}")
+
+    # Plot the results for all trials
+    plot_results(result_grid, metric=args.metric, filename="ray-loss-plot.png")
+    plot_results(result_grid, metric="train_loss",
+                 filename="ray-train_loss-plot.png")
+
+
+def plot_results(result_grid, metric="loss", filename="plot.png"):
+    """
+    Plot the results for all trials and save the plot to a file.
+
+    Args:
+    - result_grid: Results from Ray Tune trials.
+    - metric: The metric to plot (e.g., 'loss').
+    - filename: Name of the file to save the plot.
+    """
+    ax = None
+    for result in result_grid:
+        label = f"lr={result.config['lr']:.6f}, batch size={result.config['batch_size']}"
+        if ax is None:
+            ax = result.metrics_dataframe.plot(
+                "training_iteration", metric, label=label)
+        else:
+            result.metrics_dataframe.plot(
+                "training_iteration", metric, ax=ax, label=label)
+
+    ax.set_title(
+        f"{metric.capitalize()} vs. Training Iteration for All Trials")
+    ax.set_ylabel(metric.capitalize())
+
+    # Save the plot to a file
+    plt.savefig(filename)
+
+    # Show the plot
+    plt.show()
+
+
+# Main entry point for script execution
```

Comment:
Remove unused arguments

Reply:
I am not entirely sure which lines you are referring to here